clang/runtime/dpct-rt/include/sparse_utils.hpp.inc

```diff
@@ -78,6 +79,241 @@ private:
+           const int *row_ptr, const int *col_ind, const T *b, int ldb,
+           const T *beta, T *c, int ldc) {
+  using Ty = typename dpct::DataType<T>::T2;
+  auto alpha_value =
+      detail::get_value(reinterpret_cast<const Ty *>(alpha), queue);
+  auto beta_value =
+      detail::get_value(reinterpret_cast<const Ty *>(beta), queue);
+
+  oneapi::mkl::sparse::matrix_handle_t *sparse_matrix_handle =
+      new oneapi::mkl::sparse::matrix_handle_t;
+  oneapi::mkl::sparse::init_matrix_handle(sparse_matrix_handle);
+  auto data_row_ptr = detail::get_memory(const_cast<int *>(row_ptr));
+  auto data_col_ind = detail::get_memory(const_cast<int *>(col_ind));
+  auto data_val =
+      detail::get_memory(reinterpret_cast<Ty *>(const_cast<T *>(val)));
+  oneapi::mkl::sparse::set_csr_data(queue, *sparse_matrix_handle, sparse_rows,
+                                    sparse_cols, info->get_index_base(),
+                                    data_row_ptr, data_col_ind, data_val);
+
+  auto data_b = detail::get_memory(reinterpret_cast<Ty *>(const_cast<T *>(b)));
+  auto data_c = detail::get_memory(reinterpret_cast<Ty *>(c));
+  switch (info->get_matrix_type()) {
+  case matrix_info::matrix_type::ge: {
+    oneapi::mkl::sparse::gemm(queue, oneapi::mkl::layout::row_major, trans,
+                              oneapi::mkl::transpose::nontrans, alpha_value,
+                              *sparse_matrix_handle, data_b, dense_cols, ldb,
+                              beta_value, data_c, ldc);
+    break;
+  }
+  default:
+    throw std::runtime_error(
+        "the csrmm does not support matrix_info::matrix_type::sy, "
+        "matrix_info::matrix_type::tr amd matrix_info::matrix_type::he");
+  }
+
+  sycl::event e =
+      oneapi::mkl::sparse::release_matrix_handle(queue, sparse_matrix_handle);
+  queue.submit([&](sycl::handler &cgh) {
+    cgh.depends_on(e);
+    cgh.host_task([=] { delete sparse_matrix_handle; });
+  });
+}
+// DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|optimize_info|dpct::sparse
+// DPCT_DEPENDENCY_EMPTY
+// DPCT_CODE
+/// Saving the optimization information for solving a system of linear
+/// equations.
+class optimize_info {
+public:
+  /// Constructor
+  optimize_info() { oneapi::mkl::sparse::init_matrix_handle(&_matrix_handle); }
+  /// Destructor
+  ~optimize_info() {
+    oneapi::mkl::sparse::release_matrix_handle(get_default_queue(),
+                                               &_matrix_handle, _deps)
+        .wait();
+  }
+  /// Add dependency for the destructor.
+  /// \param [in] e The event which the destructor depends on.
+  void add_dependency(sycl::event e) { _deps.push_back(e); }
+  /// Get the internal saved matrix handle.
+  /// \return Returns the matrix handle.
+  oneapi::mkl::sparse::matrix_handle_t get_matrix_handle() const noexcept {
+    return _matrix_handle;
+  }
+
+private:
+  oneapi::mkl::sparse::matrix_handle_t _matrix_handle = nullptr;
+  std::vector<sycl::event> _deps;
+};
```

Comment:
This `optimize_info` is used to migrate `cusparseSolveAnalysisInfo_t`. More precisely, the type `cusparseSolveAnalysisInfo_t` will be migrated to `std::shared_ptr<optimize_info>` because `cusparseSolveAnalysisInfo_t` is a pointer and user may assign 0 to it.

In cuSPARSE, a `cusparseSolveAnalysisInfo_t` variable is used to pass the analyzed info from `csrsv_analysis()` to `csrsv()`. Since we decide to migrate `csrsv_analysis()` to the wrapper of `optimize_trsv()` and the optimized info is saved in the `matirx_handle` variable, then we want to use `matirx_handle` as the equivalent of `cusparseSolveAnalysisInfo_t`.

In cuda, cusparseDestroySolveAnalysisInfo() is a sycn API, it does not depends on cudaStream varaible as input. But in MKL, `release_matrix_handle` need a queue as input. Although the queues in the migrated code are all in-ordered queues, but we cannot make sure whether the `default_queue` and the queue user using for `csrsv_analysis()` and `csrsv()` are the same queue. So additional dependencies need to be added for the de-constructor by the `add_dependency()` interface. See line 312, we added the event as dependency. And similar thing will be done when we add new helper function `csrsv()`.

Reply:
you could do the same things for all the optimize_xyz functions -- they and the matrix_handle can persist beyond different calls adn the optimizations housed inside of them can accumulate for a given sparsity pattern...

Reply:
I will do the similar things if necessary.
Fo example, when migrating `cusparseScsrmv`, since all information is available in this call, then we can call optimize_xyz in the migrated wrapper. But for `cusparseSpMV`, since the sparse matrix info is set by `cusparseCreateCsr`, then we need a similar structure as `cusparseSpMatDescr_t` to pass the information among different calls.